src/p384/mod.rs

```diff
@@ -0,0 +1,98 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * (c) ZeroTier, Inc.
+ * https://www.zerotier.com/
+ */
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::io::Write;
+use std::mem::{size_of, transmute};
+use std::str::FromStr;
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use crate::zeroize::{Zeroize, ZeroizeOnDrop};
+use zerotier_common_utils::{base64, blob::Blob, error::InvalidParameterError, tofrombytes::ToFromBytes};
+use zerotier_crypto_glue::{hash::SHA384, p384::*};
+
+use crate::{base24, base62};
+use crate::{ADDRESS_ERR, IDENTITY_ERR};
+
+mod address;
+mod identifier;
+mod identity;
+mod identity_secret;
+mod short_address;
+
+pub use address::*;
+pub use identifier::*;
+pub use identity::*;
+pub use identity_secret::*;
+pub use short_address::*;
+
+/// We won't have more than one identity type for a while, but putting a variant number will allow
+/// us to treat an identity as a discriminated union when we add more.
+/// For security purposes this must be a statically assigned unique number.
+const IDENTITY_VARIANT_P384: u8 = 1;
+
+const DOMAIN_MASTER_SIG: &[u8] = b"ZTID1_MASTERSIG";
+const DOMAIN_SUBKEY_SIG: &[u8] = b"ZTID1_SUBKEYSIG";
+
+const PREFIX_IDENTITY: &str = "zt:id:";
+const PREFIX_ADDRESS: &str = "zt:addr:";
+const PREFIX_SHORT: &str = "zt:addr:";
+
+fn first_128_to_string(b: &[u8], s: &mut String) {
```

Comment:
I'm a little confused about how we pick between bases 24, 62, and 64; they all seem to be used in various `to_string` variants, and it isn't immediately obvious when one should be chosen over another.

Some documentation would probably suffice -- apologies if it already exists elsewhere! -- or enforcement via some general trait/code structure, but right now it's hard to evaluate which one is appropriate for a given scenario.

Reply:
This is another instance were I didn't write the code. If I could make the github PR ignore the very first commit to this branch that would be reflected in the diff. So I don't actually know! Adam does.